Prevent hidden documents from locking the screen orientation

marcoscaceres opened the pull request:
#### a68ef76884df02cb25c760874bfc38df99c09783
<pre>
Prevent hidden documents from locking the screen orientation
<a href="https://bugs.webkit.org/show_bug.cgi?id=247248">https://bugs.webkit.org/show_bug.cgi?id=247248</a>
rdar://102019707

Reviewed by Youenn Fablet.

* LayoutTests/fast/screen-orientation/hidden-document-check-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.action_sequence):
(async if):
(window.test_driver_internal.minimize_window):
(window.test_driver_internal.set_window_rect):
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/hidden_document-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/hidden_document.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):

Canonical link: <a href="https://commits.webkit.org/257019@main">https://commits.webkit.org/257019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7728e53740054c201264599410236e60ba99348

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97575 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6826 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107073 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167338 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101528 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7174 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35584 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89971 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103738 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103225 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5369 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84205 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32375 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89039 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75299 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/825 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/813 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21971 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4832 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5630 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44440 "Found 1 new test failure: fast/images/animated-heics-draw.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41357 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/page/ScreenOrientation.cpp:
```diff
@@ -154,6 +160,9 @@ ExceptionOr<void> ScreenOrientation::unlock()
         return Exception { InvalidStateError, "Document is not fully active."_s };
 
+    if (document->page() && !document->page()->isVisible())
+        return Exception { SecurityError, "Only visible documents can unlock the screen orientation"_s };
```
If the document did lock the screen, shouldn't it be able to unlock it whenever it wants?
Or maybe, it is somehow not possible, in which case do we need that check?
If we want that check, should it be SecurityError or InvalidStateError?
Also we are doing this check before the isSameOriginAsTopDocument check in unlock but after in lock. Should we be consistent?

marcoscaceres replied:
> If the document did lock the screen, shouldn't it be able to unlock it whenever it wants? Or maybe, it is somehow not possible, in which case do we need that check?

I was thinking of a case where the document goes fullscreen, switches orientation, then the user switches to another application (but the web app keeps running in the background).

It could unlock, but then we would need to hold that state and do the right thing when it's visible.

> If we want that check, should it be SecurityError or InvalidStateError?

It felt like a `SecurityError`, as I was treating it as a "pre-lock condition", which are `SecurityError`.  

> Also we are doing this check before the isSameOriginAsTopDocument check in unlock but after in lock. Should we be consistent?

Oops, yes. Will fix. 

As an aside, the `isSameOriginAsTopDocument` will eventually go away. That shouldn't be there. But it does need a different check, so should be consistent.

youennf commented on Source/WebCore/page/ScreenOrientation.cpp:
```diff
@@ -110,6 +110,12 @@ void ScreenOrientation::lock(LockType lockType, Ref<DeferredPromise>&& promise)
+        promise->reject(Exception { SecurityError, "Only visible documents can lock the screen orientation"_s });
+        return;
+    }
+
```
Unrelated to this patch, but we could try to use DOMPromiseDeferred::settle in manager->lock().
Maybe by adding settle(std::optional<Exception>) to it?

youennf commented on LayoutTests/TestExpectations:
```diff
@@ -6041,3 +6041,7 @@ http/tests/loading/oauth.html [ DumpJSConsoleLogInStdErr ]
 imported/w3c/web-platform-tests/html/infrastructure/urls/terminology-0/document-base-url-initiated-grand-parent.https.window.html [ DumpJSConsoleLogInStdErr ]
+
+# Screen orientation
+## We don't implment test_driver.minimize_window()
```
Is `window.internals.setPageVisibility` good enough to implement test_driver.minimize_window?
If so, could we try implementing it this way?

marcoscaceres replied:
I had a look. We might be able to use the same semantics ((ab)using page visibility), but I don't think we can use `window.internal`, as it needs to be done through `TestDriver`. 

Is my understanding correct?

youennf replied:
Can we update  LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js to use window.internals?

marcoscaceres replied:
Indeed we can, I recently discovered. 🎉 I’ll patch something up.

youennf commented on LayoutTests/fast/screen-orientation/hidden-document-check.html:
```diff
@@ -0,0 +1,35 @@
+      var jsTestIsAsync = true;
+
+      async function startTest() {
+        window.internals?.setPageVisibility(false);
```
I did not know `?`. Is it equivalent to if (window.internals) window.internals.set... ?

marcoscaceres replied:
I guess the right thing to do should be to stop the test if `window.internals` is not there. There is no point in proceeding if we can't set the page visibility.

youennf replied:
Either that or if you want to allow manual testing as well in Safari, make it clear through the web page to the manual tester what to do.

marcoscaceres replied:
I went with forcing an error. These tests really shouldn't be run manually.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -338,3 +338,52 @@ window.test_driver_internal.consume_user_activation = async function(context)
+            resolve(rect);
+        }, { once: true });
+        testRunner.setPageVisibility("hidden");
+    });
```
Should be `await` unless you meant to return a Promise wrapped inside a Promise.

Or to preserve current behavior, but return undefined, you can just omit `return` from the line

marcoscaceres replied:
Yeah, good suggestion... the method still needs to return the DOMRect, but I can simplify it.

youennf replied:
`return` is fine but I would keep the style of set_window_rect and minimize_window consistent (await/return and line breaks).

youennf commented on LayoutTests/fast/screen-orientation/hidden-document-check.html:
```diff
@@ -0,0 +1,35 @@
+      var jsTestIsAsync = true;
+
+      async function startTest() {
+        window.internals.setPageVisibility(false);
```
Is this test now redundant with below WPT test?

marcoscaceres replied:
Yeah, I guess it is. I'll delete it.

marcoscaceres committed:
Prevent hidden documents from locking the screen orientation
https://bugs.webkit.org/show_bug.cgi?id=247248
rdar://102019707

Reviewed by Youenn Fablet.

* LayoutTests/fast/screen-orientation/hidden-document-check-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.action_sequence):
(async if):
(window.test_driver_internal.minimize_window):
(window.test_driver_internal.set_window_rect):
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/hidden_document-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/hidden_document.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):

Canonical link: https://commits.webkit.org/257019@main
